src/java/org/apache/cassandra/db/compaction/CompactionThroughputThrottler.java

```diff
@@ -0,0 +1,71 @@
+package org.apache.cassandra.db.compaction;
+
+import com.google.common.util.concurrent.RateLimiter;
+import com.palantir.logsafe.SafeArg;
+import org.apache.cassandra.db.ColumnFamilyStore;
+import org.apache.cassandra.db.Directories;
+import org.apache.cassandra.service.StorageService;
+import org.apache.cassandra.utils.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Backpressure mechanism to reduce compaction throughput for large tables as disk fills up
+ */
+public class CompactionThroughputThrottler
+{
+    private static final boolean ENABLE_COMPACTION_THROUGHPUT_THROTTLING = Boolean.getBoolean("palantir_cassandra.enable_compaction_throughput_throttling");
+    private static final long THROTTLER_TABLE_SIZE_BYTES = Long.getLong("palantir_cassandra.throttler_table_size_gb", 10L) * 1024L * 1024L * 1024L;
+    private static final Logger logger = LoggerFactory.getLogger(CompactionThroughputThrottler.class);
+
+    public static RateLimiter getRateLimiter(Pair<String, String> ksCf)
+    {
+        String keyspace = ksCf.left;
+        String columnFamily = ksCf.right;
+        RateLimiter defaultRateLimit = RateLimiter.create(CompactionManager.instance.getRateLimiter().getRate());
+        if (!ENABLE_COMPACTION_THROUGHPUT_THROTTLING)
+        {
+            return defaultRateLimit;
+        }
+
+        double diskUsage = Directories.getMaxPathToUtilization().getValue();
+        long liveSpaceUsedBytes = ColumnFamilyStore.getIfExists(keyspace, columnFamily).metric.liveDiskSpaceUsed.getCount();
```

Comment:
What happens if the ksCf pair doesn't exist here? Do we need to add error handling?

Reply:
We're not getting this far if the ks / cf doesn't exist. Happy to add extra try / catch or a pre-condition but it feels redundant.

Reply:
It will NPE

```
java.lang.NullPointerException
	at com.palantir.cassandra.db.compaction.CompactionThroughputThrottler.getRateLimiter(CompactionThroughputThrottler.java:34)
	at org.apache.cassandra.db.compaction.LeveledCompactionStrategy$LeveledScanner.<init>(LeveledCompactionStrategy.java:338)
	at org.apache.cassandra.db.compaction.LeveledCompactionStrategy.getScanners(LeveledCompactionStrategy.java:265)
	at org.apache.cassandra.db.compaction.WrappingCompactionStrategy.getScanners(WrappingCompactionStrategy.java:458)
	at org.apache.cassandra.db.compaction.AbstractCompactionStrategy.getScanners(AbstractCompactionStrategy.java:355)
	at org.apache.cassandra.db.compaction.CompactionTask.runMayThrow(CompactionTask.java:174)
	at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:28)
	at org.apache.cassandra.db.compaction.CompactionTask.executeInternal(CompactionTask.java:76)
	at org.apache.cassandra.db.compaction.AbstractCompactionTask.execute(AbstractCompactionTask.java:59)
	at org.apache.cassandra.db.compaction.CompactionManager$10.runMayThrow(CompactionManager.java:746)
	at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:28)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

Since this goes through [CompactionTask#runMayThrow](https://github.com/palantir/cassandra/blob/palantir-cassandra-2.2.18/src/java/org/apache/cassandra/db/compaction/CompactionTask.java#L102) we use a [try with resources](https://github.com/palantir/cassandra/blob/palantir-cassandra-2.2.18/src/java/org/apache/cassandra/db/compaction/CompactionTask.java#L173-L174) for the table refs - which will be auto-closed when we error out. It's better for us to propagate the exception than to try to handle it.